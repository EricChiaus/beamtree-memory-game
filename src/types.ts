export interface FlippableObject {
    isFlippable: boolean;
    isFlipped: boolean;
}

export interface FlippableCard extends FlippableObject {
    index: number;
    content: string;
}

export interface Game<T> {
    items: T[];
    isDisabled: boolean;
    hasCompleted: boolean;
    restart: () => void;
    flip: (index: number) => void;
    getItem: (index: number) => T | null;
}
