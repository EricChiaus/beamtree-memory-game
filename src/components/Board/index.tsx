import React, { useCallback, useContext } from 'react';
import { BoardWrapper } from './index.style';
import GameContext from './../../contexts/GameContext';
import Flippable from '../Flippable';
import Card from '../Card';
import { FlippableCard } from '../../types';

const Board: React.FC = () => {
    const { items, flip } = useContext(GameContext);

    const onFlip = useCallback(
        (item: FlippableCard) => () => {
            flip(item.index);
        },
        [flip],
    );

    return (
        <BoardWrapper>
            {items.map(item => (
                <Flippable
                    key={item.index}
                    dataTestId={`flippable-card-${item.index}`}
                    isFlipped={item.isFlipped}
                    onFlip={onFlip(item)}>
                    <Card index={item.index} />
                </Flippable>
            ))}
        </BoardWrapper>
    );
};

export default Board;
