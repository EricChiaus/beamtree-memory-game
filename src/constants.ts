export const BOARD_WIDTH = 4;
export const BOARD_HEIGHT = 4;
export const CARDS_NUMBER = BOARD_WIDTH * BOARD_HEIGHT;
