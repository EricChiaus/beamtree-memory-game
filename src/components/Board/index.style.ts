import styled from 'styled-components';
import { BOARD_HEIGHT, BOARD_WIDTH } from './../../constants';

export const BoardWrapper = styled.div`
    display: grid;
    grid-template-columns: ${() =>
        new Array(BOARD_WIDTH).fill('148px').join(' ')};
    grid-template-rows: ${() =>
        new Array(BOARD_HEIGHT).fill('198px').join(' ')};
`;
