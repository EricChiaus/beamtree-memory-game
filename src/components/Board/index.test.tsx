import React, { useContext } from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { FlippableCard } from '../../types';
import { CARDS_NUMBER } from '../../constants';
import Board from '.';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(),
}));
const mockUseContext = useContext as jest.Mock;

jest.mock('./index.style', () => ({
    BoardWrapper: ({ children }) => <div>{children}</div>,
}));

jest.mock('../Flippable', () => ({ dataTestId, isFlipped, onFlip }) => (
    <button data-testid={dataTestId} onClick={onFlip}>
        {isFlipped}
    </button>
));

const generateCards = (): FlippableCard[] => {
    const cards: FlippableCard[] = [];
    for (let i = 0; i < CARDS_NUMBER; i++) {
        cards.push({
            index: i,
            content: Math.floor(i / 2).toString(),
            isFlippable: true,
            isFlipped: true,
        });
    }
    return cards;
};

const flipSpy = jest.fn();

describe('<Board />', () => {
    beforeEach(() => {
        mockUseContext.mockReset().mockReturnValue({
            items: generateCards(),
            flip: flipSpy,
        });
    });

    it('should match the snapshot', () => {
        const { container } = render(<Board />);
        expect(container).toMatchSnapshot();
    });

    it('should invoke onFlip event on Flippable components', () => {
        render(<Board />);
        fireEvent.click(screen.getByTestId('flippable-card-0'));
        expect(flipSpy).toBeCalledWith(0);
    });
});
