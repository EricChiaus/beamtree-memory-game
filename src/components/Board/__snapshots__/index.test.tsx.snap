// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> should match the snapshot 1`] = `
<div>
  <div>
    <button
      data-testid="flippable-card-0"
    />
    <button
      data-testid="flippable-card-1"
    />
    <button
      data-testid="flippable-card-2"
    />
    <button
      data-testid="flippable-card-3"
    />
    <button
      data-testid="flippable-card-4"
    />
    <button
      data-testid="flippable-card-5"
    />
    <button
      data-testid="flippable-card-6"
    />
    <button
      data-testid="flippable-card-7"
    />
    <button
      data-testid="flippable-card-8"
    />
    <button
      data-testid="flippable-card-9"
    />
    <button
      data-testid="flippable-card-10"
    />
    <button
      data-testid="flippable-card-11"
    />
    <button
      data-testid="flippable-card-12"
    />
    <button
      data-testid="flippable-card-13"
    />
    <button
      data-testid="flippable-card-14"
    />
    <button
      data-testid="flippable-card-15"
    />
  </div>
</div>
`;
